projects/sandbox/train/train/data/base.py

```diff
@@ -0,0 +1,516 @@
+                return device_ids
+            else:
+                device_id = self.trainer.device_ids[rank]
+                return f"cuda:{device_id}"
+
+    def get_world_size_and_rank(self) -> tuple[int, int]:
+        """
+        Name says it all, but generalizes to the case
+        where we aren't running distributed training.
+        """
+        if not torch.distributed.is_initialized():
+            return 1, 0
+        else:
+            world_size = torch.distributed.get_world_size()
+            rank = torch.distributed.get_rank()
+            return world_size, rank
+
+    def get_logger(self, world_size, rank):
+        logger_name = "PEDataset"
+        if world_size > 1:
+            logger_name += f":{rank}"
+        return logging.getLogger(logger_name)
+
+    # ================================================ #
+    # Re-paramterizing some attributes
+    # ================================================ #
+
+    def calc_sample_length(self) -> float:
+        """Length of samples generated by datasets in seconds"""
+        return (
+            self.hparams.kernel_length
+            + self.hparams.fduration
+            + self.hparams.psd_length
+        )
+
+    @property
+    def num_workers(self):
+        local_world_size = len(self.trainer.device_ids)
+        return min(6, int(os.cpu_count() / local_world_size))
+
+    @property
+    def pre_whiten_size(self):
+        """Size of kernel before whitening filter is cut off"""
+        length = self.hparams.kernel_length + self.hparams.fduration
+        return int(length * self.sample_rate)
+
+    @property
+    def strain_dim(self):
+        """
+        Size of strain tensor timeseries
+        the nerual network will analyze
+        """
+        return int(self.hparams.kernel_length * self.sample_rate)
+
+    @property
+    def num_params(self):
+        """Number of parameters inference is performed on"""
+        return len(self.hparams.inference_params)
+
+    @property
+    def val_batch_size(self):
+        """Use larger batch sizes when we don't need gradients."""
+        return int(4 * self.hparams.batch_size)
+
+    @property
+    def train_val_fnames(self):
+        """List of background files used for both training and validation"""
+        background_dir = self.data_dir / "train" / "background"
+        fnames = list(background_dir.glob("*.hdf5"))
+        return fnames
+
+    @property
+    def test_fnames(self):
+        """List of background files used for testing a trained model"""
+        test_dir = self.data_dir / "test" / "background"
+        fnames = list(test_dir.glob("*.hdf5"))
+        return fnames
+
+    def train_val_split(self) -> Sequence[str]:
+        """
+        Split background files into training and validation sets
+        based on the requested duration of the validation set
+        """
+        fnames = sorted(self.train_val_fnames)
+        durations = [int(fname.stem.split("-")[-1]) for fname in fnames]
```

Comment:
This is making some assumptions about the filenames that did not used to be the case before. Use a regex pattern here to be more robust and explicit. Also, mention in a comment how what is the standard format, for example, gwosc hdf5 format.

Reply:
Yeah I can add a regex for it: The format is `{prefix}-{t0}-{duration}.hdf5`.

Reply:
I was trying this out with some gwosc background, and now thinking about this further, this also needs the ifo in front. Like `H-H1_GWOSC_O3b_4KHZ_R1-1262243840-4096.hdf5` ; `L-L1_GWOSC_O3b_4KHZ_R1-1262243840-4096.hdf5`. Same under /dev/shm/kafka: `H-H1_O3ReplayMDC_llhoft-1394823647-1.gwf`

Reply:
What I am saying is we should move to something like the online streaming frames instead of using something custom.